packetbeat/protos/rpc/rpc.go

```diff
@@ -0,0 +1,233 @@
+	if !testMode {
+		if err := cfg.Unpack(&config); err != nil {
+			logp.Warn("failed to read config")
+			return nil, err
+		}
+	}
+
+	if err := p.init(results, &config); err != nil {
+		logp.Warn("failed to init")
+		return nil, err
+	}
+	return p, nil
+}
+
+func (rpc *Rpc) init(results publish.Transactions, config *rpcConfig) error {
+	rpc.setFromConfig(config)
+	rpc.results = results
+
+	return nil
+}
+
+func (rpc *Rpc) setFromConfig(config *rpcConfig) error {
+	rpc.Ports = config.Ports
+	rpc.transactionTimeout = time.Duration(config.TransactionTimeout) * time.Second
+	return nil
+}
+
+func (rpc *Rpc) GetPorts() []int {
+	return rpc.Ports
+}
+
+// Called when TCP payload data is available for parsing.
+func (rpc *Rpc) Parse(
+	pkt *protos.Packet,
+	tcptuple *common.TcpTuple,
+	dir uint8,
+	private protos.ProtocolData,
+) protos.ProtocolData {
+
+	defer logp.Recover("ParseRPC exception")
+
+	conn := ensureRpcConnection(private)
+
+	conn = rpc.handleRpcFragment(conn, pkt, tcptuple, dir)
+	if conn == nil {
+		return nil
+	}
+	return conn
+}
+
+// Called when the FIN flag is seen in the TCP stream.
+func (rpc *Rpc) ReceivedFin(tcptuple *common.TcpTuple, dir uint8,
+	private protos.ProtocolData) protos.ProtocolData {
+
+	defer logp.Recover("ReceivedFinRpc exception")
+
+	// forced by TCP interface
+
+	// TODO
+	return private
+}
+
+// Called when a packets are missing from the tcp
+// stream.
+func (rpc *Rpc) GapInStream(tcptuple *common.TcpTuple, dir uint8,
+	nbytes int, private protos.ProtocolData) (priv protos.ProtocolData, drop bool) {
+
+	defer logp.Recover("GapInRpcStream exception")
+
+	// forced by TCP interface
+
+	// TODO
+	return private, false
+}
+
+// ConnectionTimeout returns the per stream connection timeout.
+// Return <=0 to set default tcp module transaction timeout.
+func (rpc *Rpc) ConnectionTimeout() time.Duration {
+	// forced by TCP interface
+	return rpc.transactionTimeout
+}
+
+func ensureRpcConnection(private protos.ProtocolData) *rpcConnectionData {
+	conn := getRpcConnection(private)
+	if conn == nil {
+		conn = &rpcConnectionData{}
+	}
+	return conn
+}
+
+func getRpcConnection(private protos.ProtocolData) *rpcConnectionData {
+	if private == nil {
+		return nil
+	}
+
+	priv, ok := private.(*rpcConnectionData)
+	if !ok {
+		logp.Warn("rpc connection data type error")
+		return nil
+	}
+	if priv == nil {
+		logp.Warn("Unexpected: rpc connection data not set")
+		return nil
+	}
+
+	return priv
+}
+
+// Parse function is used to process TCP payloads.
+func (rpc *Rpc) handleRpcFragment(
+	conn *rpcConnectionData,
+	pkt *protos.Packet,
+	tcptuple *common.TcpTuple,
+	dir uint8,
+) *rpcConnectionData {
+
+	st := conn.Streams[dir]
+	if st == nil {
+		st = newStream(pkt, tcptuple)
+		conn.Streams[dir] = st
+	} else {
+		// concatenate bytes
+		st.rawData = append(st.rawData, pkt.Payload...)
+		if len(st.rawData) > tcp.TCP_MAX_DATA_IN_STREAM {
+			logp.Debug("rpc", "Stream data too large, dropping TCP stream")
```

Comment:
I would recommend creating a single `debugf` function for your package. And everywhere you write `logp.Debug("rpc", "some message")` replace it with `debugf("some message")`. It makes it easy to see what the debug selectors are for the package and it avoids the problem of typos with the selector.

Basically at the package level just add this:

`var debugf = logp.MakeDebug("rpc")`